File: clojure/tp/maps.clj
(ns clojure.example.maps
  (:require [clojure.set])
  (:gen-class))

(defn clojure-maps []
  (println "clojure comes with built-in HashMaps and SortedMaps")

  ; hash-map 
  (def plain-hashmap (hash-map "year" "2022" "month" "january" "day" "10"))
  (println plain-hashmap)

  ; sorted-map
  (def sorted-hashmap (sorted-map "year" "2022" "month" "january" "day" "10"))
  (println sorted-hashmap)

  ; get to retrieve value from provided key
  (println (get plain-hashmap "year"))

  ; contains? checks if key exists in the hashmap
  (println (contains? plain-hashmap "day"))
  (println (contains? plain-hashmap "time"))

  ; find returns the [key value] pair for provided key
  (println (find plain-hashmap "month"))
  (println (find plain-hashmap "time"))

  ; keys returns a list of keys in the map
  (println (keys plain-hashmap))

  ; vals returns a list of values in the map
  (println (vals plain-hashmap))

  ; dissoc returns a new map with a key-value removed
  (println (dissoc plain-hashmap "year"))

  ; merge merges two maps
  (def hour-map (hash-map "hour" "22" "minute" "27" "second" "55"))
  (println (merge plain-hashmap hour-map))

  ; merge-with merges two maps, appling an operation for keys that exist
  ; in both maps (like sum, subtraction, etc)
  (def a (hash-map :x 10 :y 20 :z 666))
  (def b (hash-map :x  9 :y 21))
  (println "merging two maps, subtracting value that exists in both")
  (println "map a" a)
  (println "map b" b)
  (println "merged with subtraction: " (merge-with - a b))

  ; select-keys returns a map only with the selected entries
  (println (select-keys plain-hashmap ["day" "month"]))

  ; rename-keys returns a new map with the renamed keys
  (println (clojure.set/rename-keys plain-hashmap {"year" "current-year"}))

  ; map-invert transform keys in values, and values in keys
  (println (clojure.set/map-invert plain-hashmap)))

(clojure-maps)

File: erlang/tp/maps_.erl

-module(maps_).
-export([start/0]).
-on_load(start/0).

start() ->
  % a simple map
  MeMap = #{ name => eduardo, age => 34 },
  io:fwrite("~p~n", [map_size(MeMap)]),

  % an example list
  DateList = [{"year", 2022}, {"month", 4}, {"day", 12}],
  DateMap = maps:from_list(DateList),

  io:fwrite("map from list ~p~n", [ DateMap ]),

  % find returns a tuple {ok, Value} if key:value exist
  % it errors otherwise
  io:fwrite("find to check keys exists ~p~n", [ maps:find("year", DateMap) ]),

  io:fwrite("get value from key ~p~n", [ maps:get("year", DateMap) ]),

  % just a boolean check if key exists (different from maps:find)
  io:fwrite("is_key checks if key exist ~p~n", [maps:is_key("year", DateMap)]),

  io:fwrite("keys gets all keys ~p~n", [maps:keys(DateMap)  ]),
  io:fwrite("values gets values ~p~n", [maps:values(DateMap)]),
  
  % longer setup for maps:merge...
  %
  ListA = [ {"a", 1984}, {"b", 1789} ],
  ListB = [ {"c", 2022}, {"d", 2035} ],

  MapA = maps:from_list(ListA),
  MapB = maps:from_list(ListB),

  io:fwrite("merging maps ~p~n", [maps:merge(MapA, MapB)]),

  io:fwrite("put adds key to map ~p~n", [maps:put("hour", 20, DateMap)]),
  io:fwrite("remove removes ~p~n", [maps:remove("year", DateMap)]).

